lib/screens/v2/widget_instance/elevator_closures.ex

```diff
@@ -32,15 +25,111 @@ defmodule Screens.V2.WidgetInstance.ElevatorClosures do
           }
   end
 
+  defmodule Upcoming do
+    @moduledoc false
+
+    @enforce_keys ~w[period]a
+    defstruct @enforce_keys ++ ~w[summary]a
+
+    @type t :: %__MODULE__{period: {Date.t(), Date.t() | nil}, summary: String.t() | nil}
+
+    @nbsp "\u00A0"
+
+    @spec serialize(t(), DateTime.t()) :: map()
+    def serialize(%__MODULE__{period: {start_date, end_date}, summary: summary}, now) do
+      today = Util.service_date(now)
+      indefinite? = is_nil(end_date)
+      starts_tomorrow? = Date.diff(start_date, today) == 1
+      starts_this_week? = Date.beginning_of_week(start_date) == Date.beginning_of_week(today)
+      relative_start? = starts_tomorrow? or starts_this_week?
+      single_day? = start_date == end_date
+
+      same_month? =
+        not is_nil(end_date) and
+          {start_date.year, start_date.month} == {end_date.year, end_date.month}
+
+      {banner_title, details_titles} =
+        cond do
+          starts_tomorrow? ->
+            both_titles("Tomorrow")
+
+          starts_this_week? ->
+            both_titles("This #{day_name(start_date)}")
+
+          true ->
+            {
+              "#{day_name(start_date)}, #{short(start_date)}",
+              [
+                "#{day_name(start_date)}, #{long(start_date)}",
+                "#{day_name(start_date)}, #{short(start_date, _dot = true)}"
+              ]
+            }
+        end
```

Comment:
These long blocks make the serialize function a bit long to parse. Could the title and trailer logic be pulled out into their own functions?

```
{banner_title, details_titles} = title_for_start_date(start_date, today)
{banner_postfix, details_postfix} = postfix_for_start_period(start_date, end_date, today)
```

The postfix generation is what uses almost all of the variables you set at the beginning of the serialize method anyway.

Reply:
I'll give this a shot, definitely.

Reply:
Committed, let me know what you think!